Sources/ImperialMicrosoft/MicrosoftRouter.swift

```diff
@@ -0,0 +1,44 @@
+import Vapor
+import Foundation
+
+public class MicrosoftRouter: FederatedServiceRouter {
+        
+    public static var tenantIDEnvKey: String = "MICROSOFT_TENANT_ID"
+
+    public let tokens: FederatedServiceTokens
+    public let callbackCompletion: (Request, String)throws -> (EventLoopFuture<ResponseEncodable>)
+    public var scope: [String] = []
+    public let callbackURL: String
+    public var tenantID: String { Environment.get(MicrosoftRouter.tenantIDEnvKey) ?? "common" }
+    public var accessTokenURL: String { "https://login.microsoftonline.com/\(self.tenantID)/oauth2/v2.0/token" }
+    public let service: OAuthService = .microsoft
+    public let errorKey = "error_description"
+    
+    public required init(
+        callback: String,
+        completion: @escaping (Request, String) throws -> (EventLoopFuture<ResponseEncodable>)
+    ) throws {
+        self.tokens = try MicrosoftAuth()
+        self.callbackURL = callback
+        self.callbackCompletion = completion
+    }
+
+    public func authURL(_ request: Request) throws -> String {
+        return "https://login.microsoftonline.com/\(self.tenantID)/oauth2/v2.0/authorize?"
+            + "client_id=\(self.tokens.clientID)&"
+            + "response_type=code&"
+            + "redirect_uri=\(self.callbackURL)&"
+            + "response_mode=query&"
+            + "scope=\(scope.joined(separator: "%20"))&"
+            + "prompt=consent"
+    }
```

Comment:
(Yet again) as above in `DropboxRouter.swift`, suggest using `URLComponents` to construct the string.